### ISSUE TITLE
Type refinement failing when extending class

### ISSUE THREAD

**Issue by deecewan** [facebook/flow#5741]
tl;dr - You can see the error [here (with generics)](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgCyWA8gEYBWeAxhgDwAqAfGALxgMDc61MAhgGcBRLAFEATuLjjGLPAA8MeAHYATYRKniwAb1RgwUAJbiBGAFwjyVWrO4GBNOGsvFrNes24BfdFACuyrRGzmBKZgAUfOIA5q5iktJ0ICwAPmCa0gCUuvpgRlBgUbEAdMamGGAAhKzsyv4wMDl6BgbFcSKZMilZ9mC+3kA) or [here (without)](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgCyWA8gEYBWeAxhgDwAqAfGALxgMDc61MAhgGcBYAMpwAtngDC-IWADeAX26peg4WMkkMACzwAnGerB4AHhjwA7ACYaJ02cPmoAkHEo0MALjCWAruJkBpxgqIroUH6WtACWcJZg4lgAYlGx8QAUfPoA5j6aeNp6ho5gAD6i9kZCAJQKqGBgMVBgWbl1zo2NbXmVWroG1QI1DWDhikA).

If you have a class `B` that extends a class `A`, and you have a function that accepts the union of these classes `A | B`, it seems impossible to refine the classes, even if the class has some distinctive feature.

Say `A` is an empty class, and `B` extends `A` and adds a property, `num`, of type `number`.

```js
class A {}

class B extends A {
	num: number;
}
```

If I have a function that accepts this union, it seems impossible for me to refine to the `B` class, even though it is more specific.

```js
function test(arg: A | B) {
	if (arg.num !== null) {
		(arg: B); // <--- flow errors here, saying that `arg` may be `A` still
    }
}
```

As I type this, I feel like this is due to nominal typing of classes. In that case, is there a way for me to refine these classes?

My use-case is using `axios`, whose libdef specifies an `AxiosError` type, with additional `axios`-specific properties. The actual library itself just modifies a native `Error` object, so I can't use `instanceof` with it, because it isn't actually a different instance.

**Comment by zalishchuk** [comment]
So, what is the workaround now with covering `axios` errors?

**Comment by stereobooster** [comment]
[Here is what you can do instead](https://flow.org/try/#0MYGwhgzhAEDiCmA7eAnAlsAoilB7F08AHgC5IAmM2eBA3gFDTQkCeADvAFzQBuua5egF96oSDACyLavgA8AFQB8hUhSo580Bk1YduAcgC2LAPqoa+4fQBmAV0TASaXImbwIJABRgUAc25SMiiyAFTKAD5wSKgYQQCUWozQaNbQ3n4AdLrw0AC8+dBGpub4+gnaTGk+-tCBGsFhcQDcSSJCQA). I guess the issue here is that Error class is open. E.g. you can do

```ts
const e = new Error('abc');
e.arg = 'something';
```

so runtime behaviour will be different from static one

**Comment by SamChou19815** [comment]
This kind of refinement with classes will not be supported, since classes can be extended
